google/auth/crypt/es.py

```diff
@@ -0,0 +1,208 @@
+        # First convert (r||s) raw signature to ASN1 encoded signature.
+        sig_bytes = _helpers.to_bytes(signature)
+        if len(sig_bytes) != self._r_s_size * 2:
+            return False
+        r = (
+            int.from_bytes(sig_bytes[: self._r_s_size], byteorder="big")
+            if _helpers.is_python_3()
+            else utils.int_from_bytes(sig_bytes[: self._r_s_size], byteorder="big")
+        )
+        s = (
+            int.from_bytes(sig_bytes[self._r_s_size :], byteorder="big")
+            if _helpers.is_python_3()
+            else utils.int_from_bytes(sig_bytes[self._r_s_size :], byteorder="big")
+        )
+        asn1_sig = encode_dss_signature(r, s)
+
+        message = _helpers.to_bytes(message)
+        try:
+            self._pubkey.verify(asn1_sig, message, ec.ECDSA(self._sha_algo))
+            return True
+        except (ValueError, cryptography.exceptions.InvalidSignature):
+            return False
+
+    @classmethod
+    def from_string(cls, public_key):
+        """Construct an Verifier instance from a public key or public
+        certificate string.
+
+        Args:
+            public_key (Union[str, bytes]): The public key in PEM format or the
+                x509 public key certificate.
+
+        Returns:
+            Verifier: The constructed verifier.
+
+        Raises:
+            ValueError: If the public key can't be parsed.
+        """
+        public_key_data = _helpers.to_bytes(public_key)
+
+        if _CERTIFICATE_MARKER in public_key_data:
+            cert = cryptography.x509.load_pem_x509_certificate(
+                public_key_data, _BACKEND
+            )
+            pubkey = cert.public_key()
+
+        else:
+            pubkey = serialization.load_pem_public_key(public_key_data, _BACKEND)
+
+        return cls(pubkey)
+
+
+class EsSigner(base.Signer, base.FromServiceAccountMixin):
+    """Signs messages with an ECDSA private key.
+
+    Args:
+        private_key (
+                cryptography.hazmat.primitives.asymmetric.ec.ECDSAPrivateKey):
+            The private key to sign with.
+        key_id (str): Optional key ID used to identify this private key. This
+            can be useful to associate the private key with its associated
+            public key or certificate.
+    """
+
+    def __init__(self, private_key, key_id=None):
+        self._key = private_key
+        self._key_id = key_id
+        # ECDSA raw signature has (r||s) format where r,s are two
+        # integers of size 32 bytes for P-256 curve and 48 bytes
+        # for P-384 curve. For P-256 curve, we use SHA256 hash algo,
+        # and for P-384 curve we use SHA384 algo.
+        if isinstance(private_key.curve, ec.SECP384R1):
+            self._r_s_size = 48
+            self._sha_algo = hashes.SHA384()
+            self._alg = "ES384"
+        else:
+            self._r_s_size = 32
+            self._sha_algo = hashes.SHA256()
+            self._alg = "ES256"
```

Comment:
These attributes are hard-coded in multiple places. Maybe it would be good to create a dataclass or named tuple to hold these? Something like:

```
@dataclass
class ESAttrs:
    rs_size: int
    sha_algo: hashes.HashAlgorithm
    name: str

    @classmethod
    def from_key(cls, key):
        if isinstance(private_key.curve, ec.SECP384R1):
            return cls(48, hashes.SHA384(), "ES384")
        else:
            # default to ES256
            return cls(32, hashes.SHA256(), "ES256")

class EsSigner:
    def __init__(self, private_key, key_id=None):
        self._es_attrs = ESAttrs.from_key(private_key)
```